libraries/c_sdk/standard/https/src/iot_https_client.c

```diff
@@ -0,0 +1,3362 @@
+    char pAlpnProtos[ IOT_HTTPS_MAX_ALPN_PROTOCOLS_LENGTH + 1 ] = { 0 };
+
+    /* The maximum string length of the Server host name is configured in IOT_HTTPS_MAX_HOST_NAME_LENGTH.
+     * This +1 is for the NULL terminator needed by IotNetworkServerInfo_t.pHostName. */
+    char pHostName[ IOT_HTTPS_MAX_HOST_NAME_LENGTH + 1 ] = { 0 };
+    bool reqQMutexCreated = false;
+    bool respQMutexCreated = false;
+    bool disconnectMutexCreated = false;
+    IotNetworkServerInfo_t networkServerInfo = { 0 };
+    IotNetworkCredentials_t networkCredentials = { 0 };
+    _httpsConnection_t * pHttpsConnection = NULL;
+    IotNetworkCredentials_t * pNetworkCredentials = NULL;
+
+    HTTPS_ON_NULL_ARG_GOTO_CLEANUP( pConnInfo->userBuffer.pBuffer );
+    HTTPS_ON_NULL_ARG_GOTO_CLEANUP( pConnInfo->pNetworkInterface );
+    HTTPS_ON_NULL_ARG_GOTO_CLEANUP( pConnInfo->pAddress );
+    HTTPS_ON_ARG_ERROR_GOTO_CLEANUP( pConnInfo->addressLen > 0 );
+
+    /* Make sure the connection context can fit in the user buffer. */
+    HTTPS_ON_ARG_ERROR_MSG_GOTO_CLEANUP( pConnInfo->userBuffer.bufferLen >= connectionUserBufferMinimumSize,
+                                         IOT_HTTPS_INSUFFICIENT_MEMORY,
+                                         "Buffer size is too small to initialize the connection context. User buffer size: %d, required minimum size; %d.",
+                                         ( *pConnInfo ).userBuffer.bufferLen,
+                                         connectionUserBufferMinimumSize );
+
+    /* Make sure that the server address does not exceed the maximum permitted length. */
+    HTTPS_ON_ARG_ERROR_MSG_GOTO_CLEANUP( pConnInfo->addressLen <= IOT_HTTPS_MAX_HOST_NAME_LENGTH,
+                                         IOT_HTTPS_INVALID_PARAMETER,
+                                         "IotHttpsConnectionInfo_t.addressLen has a host name length %d that exceeds maximum length %d.",
+                                         pConnInfo->addressLen,
+                                         IOT_HTTPS_MAX_HOST_NAME_LENGTH );
+
+    /* Make sure that the ALPN protocols does not exceed the maximum permitted length. */
+    HTTPS_ON_ARG_ERROR_MSG_GOTO_CLEANUP( pConnInfo->alpnProtocolsLen <= IOT_HTTPS_MAX_ALPN_PROTOCOLS_LENGTH,
+                                         IOT_HTTPS_INVALID_PARAMETER,
+                                         "IotHttpsConnectionInfo_t.alpnProtocolsLen of %d exceeds the configured maximum protocol length %d. See IOT_HTTPS_MAX_ALPN_PROTOCOLS_LENGTH for more information.",
+                                         pConnInfo->alpnProtocolsLen,
+                                         IOT_HTTPS_MAX_ALPN_PROTOCOLS_LENGTH );
+
+    pHttpsConnection = ( _httpsConnection_t * ) ( pConnInfo->userBuffer.pBuffer );
+
+    /* Start with the disconnected state. */
+    pHttpsConnection->isConnected = false;
+
+    /* Initialize disconnection state keeper. */
+    pHttpsConnection->isDestroyed = false;
+
+    /* Init
```

Comment:
At this point we'll be holding three locks. That's a design that's tough to thoroughly test, maintain, and support. Could we instead use a single mutex to protect all fields on the context handle structure?

Reply:
Yes changing.